File: java/singleton/exercise/SessionManager.java
package singleton.exercise;

import singleton.Logger;

import java.util.HashMap;
import java.util.Map;

public class SessionManager {
    private static SessionManager instance;
    private Map<String, String> userSessions;

    private SessionManager() {
        // Private constructor to prevent instantiation from outside
        // Initialize user sessions map
        this.userSessions = new HashMap<>();
    }

    public static SessionManager getInstance() {
        if (instance == null) {
            synchronized (Logger.class) {
                if(instance == null) {
                    instance = new SessionManager();
                }
            }
        }
        return instance;
    }

    public String createSession(String userId) {
        // Generate a session ID and associate it with the user
        String sessionId = generateSessionId();
        userSessions.put(sessionId, userId);
        return sessionId;
    }

    public String getUserId(String sessionId) {
        // Retrieve the user ID associated with the session
        return userSessions.get(sessionId);
    }

    private String generateSessionId() {
        // Simple session ID generation logic (for demonstration purposes)
        return "SESSION_" + System.currentTimeMillis();
    }
}

File: C#/singleton/exercise/SessionManager.cs
namespace Singleton.Exercise;

public sealed class SessionManager
{
    private static SessionManager? s_instance;
    private readonly Dictionary<string, string> _userSessions;

    private SessionManager()
    {
        _userSessions = [];
    }

    public static SessionManager GetInstance()
    {
        if (s_instance is null)
        {
            lock (typeof(SessionManager))
            {
                s_instance ??= new SessionManager();
            }
        }

        return s_instance;
    }

    public string CreateSession(string userId)
    {
        string sessionId = GenerateSessionId();
        _userSessions[sessionId] = userId;
        return sessionId;
    }

    public string? GetUserId(string sessionId)
    {
        return _userSessions.GetValueOrDefault(sessionId);
    }

    private string GenerateSessionId()
    {
        return $"SESSION_{DateTimeOffset.Now.ToUnixTimeMilliseconds()}";
    }
}
